### ISSUE TITLE
Matern kernels have serious numerical issues for multivariate problems

### ISSUE THREAD

**Issue by jmsull** [dfm/tinygp#115]
Hi, and thanks for this super clear GP package!

I tried to adapt the docs [Multivariate Data example](https://tinygp.readthedocs.io/en/stable/tutorials/multivariate.html) and wanted to access the gradient of the conditioned tinygp mean.
In doing this, I found that jax.grad() applied to the (wrapped) tinygp mean function returns NaNs when applied to a point used in training. 

Here is [a MWE gist](https://gist.github.com/jmsull/c0096588a47ba382681bf3b23850b08a) (with versions in a comment): 
the only line I changed from the docs example is 32 to line 33 - when I use the SquaredExponential kernel in line 32 as in the docs, the jax.grad returns some non-NaN floats just fine.

Also, perturbing the training point slightly (by adding 1e-6) returns a non-NaN gradient as well.

Am I missing anything here or is this a bug?

Thanks in advance,

Jamie

**Comment by dfm** [comment]
Very interesting! There was a tiny bug in your example (which should have thrown an error! I'll open a different issue), but I can reproduce the issue. There must be a divide by zero happening somewhere when dX goes to zero - let me do some digging. Thanks for the report!

**Comment by jmsull** [comment]
Ah sorry I wasn't careful about the X[0] shape. Yeah, before posting here I tried to call just the kernel with both arguments X1,X2 being the same to see if there was an issue when dX=0 but wasn't seeing any problem there - thanks for looking into this!

**Comment by dfm** [comment]
> when I use the SquaredExponential kernel in line 32 as in the docs, the jax.grad returns some non-NaN floats just fine

I was digging a little more this morning, and hadn't noticed this before! One other thing to note is that the `jaxopt` optimizer also doesn't change the parameters at all, so it looks like there's a deeper problem with the Matern kernel here. I've seen this in a few other spots too! I'm going to update the issue title to emphasize the Matern part rather than the grad part, because I think that's a bit of a red herring here...

**Comment by dfm** [comment]
@jmsull — After merging #121, your test case now passes. The issue there was that the gradient of the `L2Distance` at `X1 == X2` was NaN, so I added a special case to fix that. There remains another issue that the `L1Distance` is often numerically unstable, but I'm going to address that in a different issue. Thanks again for the report!

**Comment by jmsull** [comment]
Great news - thanks again!
